### PR TITLE
fix(interimElement): support scope.$destroy events

### LINKED ISSUE #3741: md-menu: md-backdrop does not get removed
I have a hard time reproducing this but sometimes (like every 5th time) when using the `md-menu` the `md-backdrop` does not get removed when clicking on a menu option, and the whole site becomes *unclickable*.

This is what the end of my html page looks like when it stops working.

```HTML
<div class="md-scroll-mask" style="z-index: 50">
    <div class="md-scroll-mask-bar"></div>
</div>
<md-backdrop class="md-menu-backdrop md-click-catcher md-default-theme">
</md-backdrop>
<div class="md-open-menu-container md-whiteframe-z2 md-default-theme md-active md-clickable"
     style="top: 58px; left: 171.515625px; -webkit-transform-origin: left top 0px;">
    <md-menu-content class="md-default-theme">
        <md-menu-item>
            <button class="md-button md-default-theme"
                    ng-transclude=""
                    ng-click="vm.onSplit('vertical')"
                    ng-disabled="vm.item.sizeX <= 1"
                    aria-label="split vertical"
                    tabindex="0"
                    aria-disabled="false">

                <md-icon class="mdi mdi-split ng-scope ng-isolate-scope md-default-theme"
                         aria-hidden="true">
                </md-icon>
                <span class="ng-binding ng-scope">Split vertical</span>
            </button>
        </md-menu-item>
        [...]
    </md-menu-content>
</div>
```

**EDIT:** At first I suspected md-scroll-mask caused this, but it apparently didn't

### LINKED ISSUE #4151: md-select leaves style and md-backdrop on the screen if closed without selecting an option
1. Open the `md-select` component
2. Instead of clicking an option to close it, click on a custom element on screen to transition away or hide the parent container

Result: 

The style applied to the `<body>` and the generated `<md-backdrop>` will remain in the DOM. The `md-backdrop` doesn't really affect things but the style left on the `<body>` causes unexpected errors with original scrollbar position.

### LINKED ISSUE #4405: Select: Back action with dropdown open
While a dropdown is opened, clicking the `BACK` button in the browser will cause the refreshed page to freeze. Nothing is clickable until a refresh of the page.

> It appears that the md-backdrop is not properly removed during $location change events.

screen shot 2015-09-01 at 8 59 53 am

screen shot 2015-09-01 at 9 00 10 am

### LINKED ISSUE #4504: problem md-backdrop can't remove after close bottom sheet
I write event ng-mouseover open bottom sheet, but when I mouseover multiple, bottom sheet show right data after I cancel bottom sheet, md-backdrop only remove last backdrop so screen freezes. How can i remove all backdrop after bottom sheet cancel. Thanks!

### PR DESCRIPTION
**Pull request by ThomasBurleson**
when navigation (eg $location) changes occur, the `scope.$destroy` event is triggered; which should remove/close/cleanup the following components: menu, select, dialog, toast, and bottomSheet.

Fixes #3741. Fixes #4405. Fixes #4504. Fixes #4151.

### CODE REVIEW

**Comment by ThomasBurleson** [comment]
@rschmukler, @robertmesserle, @jelbourn  - would you mind reviewing this?

File: src/components/menu/js/menuDirective.js [hunk 1]
```diff
@@ -190,10 +190,13 @@ function MenuDirective($mdUtil) {
     }
     mdMenuCtrl.init(menuContainer, { isInMenuBar: isInMenuBar });
 
-    scope.$on('$destroy', function() {
-      menuContainer.remove();
-      mdMenuCtrl.close();
-    });
+    scope.$on('$destroy', angular.bind(null, function(scope, ev) {
```
Comment (ThomasBurleson) [inline src/components/menu/js/menuDirective.js]: I will remove the [now] unnecessary `angular.bind()`

**Comment by ThomasBurleson** [comment]
@rschmukler - I need your help with select.spec.js#L58-72. 
*  If you run all the Karma tests, that `it(...)` fails. 
*  if you run only that test, the `iit(...)` passes. 
After 3hours, I still cannot identify why it is failing...

File: docs/config/template/index.template.html [hunk 2]
```diff
@@ -157,7 +157,7 @@ <h2 class="md-toolbar-item md-breadcrumb md-headline">
   </div>
 
 
-  <script src="https://ajax.googleapis.com/ajax/libs/angularjs/{$ doc.buildConfig.ngVersion $}/angular.min.js"></script>
+  <script src="https://ajax.googleapis.com/ajax/libs/angularjs/{$ doc.buildConfig.ngVersion $}/angular.js"></script>
```
Comment (jelbourn) [inline docs/config/template/index.template.html]: Be sure to undo this before submitting.

File: src/components/bottomSheet/bottom-sheet.js [hunk 3]
```diff
@@ -12,9 +12,17 @@ angular
   .directive('mdBottomSheet', MdBottomSheetDirective)
   .provider('$mdBottomSheet', MdBottomSheetProvider);
 
-function MdBottomSheetDirective() {
+/* @ngInject */
+function MdBottomSheetDirective($mdBottomSheet) {
   return {
-    restrict: 'E'
+    restrict: 'E',
+    link : function postLink(scope, element, attr, ctrl) {
```
Comment (jelbourn) [inline src/components/bottomSheet/bottom-sheet.js]: I would omit `ctrl` for a directive that doesn't have one.

File: src/components/dialog/dialog.js [hunk 4]
```diff
@@ -25,9 +25,19 @@ function MdDialogDirective($$rAF, $mdTheming) {
           //-- delayed image loading may impact scroll height, check after images are loaded
           angular.element(images).on('load', addOverflowClass);
         }
+
+        scope.$on('$destroy', angular.bind(null, function(scope, ev) {
```
Comment (jelbourn) [inline src/components/dialog/dialog.js]: I don't see what this `angular.bind` call is doing.

File: src/components/dialog/dialog.js [hunk 5]
```diff
@@ -530,16 +540,30 @@ function MdDialogProvider($$interimElementProvider) {
...
 
-      options.hideBackdrop();
+      // For navigation $destroy events, do a quick, non-animated removal,
+      // but for normal closes (from clicks, etc) animate the removal
 
-      return dialogPopOut(element, options)
-        .finally(function() {
-          angular.element($document[0].body).removeClass('md-dialog-is-showing');
-          element.remove();
+      return !!options.$destroy ? detachAndClean() : animateRemoval().then( detachAndClean );
```
Comment (jelbourn) [inline src/components/dialog/dialog.js]: The `!!` is unnecessary

File: src/components/dialog/dialog.js [hunk 6]
```diff
@@ -530,16 +540,30 @@ function MdDialogProvider($$interimElementProvider) {
...
+      /**
+       * For normal closes, animate the removal.
+       * For forced closes (like $destroy events), skip the animations
+       */
+      function animateRemoval() {
+        return dialogPopOut(element, options);
+      }
+
+      /**
+       * Detach the element
```
Comment (jelbourn) [inline src/components/dialog/dialog.js]: What does `AndClean` mean?

File: src/components/dialog/dialog.js [hunk 7]
```diff
@@ -530,16 +540,30 @@ function MdDialogProvider($$interimElementProvider) {
...
-        .finally(function() {
-          angular.element($document[0].body).removeClass('md-dialog-is-showing');
-          element.remove();
+      return !!options.$destroy ? detachAndClean() : animateRemoval().then( detachAndClean );
 
-          options.origin.focus();
-        });
+      /**
+       * For normal closes, animate the removal.
+       * For forced closes (like $destroy events), skip the animations
```
Comment (jelbourn) [inline src/components/dialog/dialog.js]: Is this explaining what this function does or how it should be used?

File: src/components/dialog/dialog.js [hunk 8]
```diff
@@ -661,10 +685,12 @@ function MdDialogProvider($$interimElementProvider) {
       /**
        * Hide modal backdrop element...
        */
-      options.hideBackdrop = function hideBackdrop() {
+      options.hideBackdrop = function hideBackdrop($destroy) {
```
Comment (jelbourn) [inline src/components/dialog/dialog.js]: It's unclear what `$destroy` is.

File: src/components/dialog/dialog.js (hunk 5 above)
Comment (jelbourn) [inline src/components/dialog/dialog.js]: Why does the option `$destroy` start with a `$`?

File: src/components/dialog/dialog.spec.js [hunk 9]
```diff
@@ -158,6 +158,27 @@ describe('$mdDialog', function() {
       expect(container.length).toBe(0);
     }));
 
+    it('should remove `md-dialog-container` on scope.$destroy()', inject(function($mdDialog, $rootScope, $timeout) {
+      var container, parent = angular.element('<div>');
+
+      $mdDialog.show(
+        $mdDialog.alert({
+          template: '<md-dialog>' +
```
Comment (jelbourn) [inline src/components/dialog/dialog.spec.js]: Can you indent the html template lines as you would standalone html?

File: src/components/select/select.js [hunk 10]
```diff
@@ -782,38 +785,40 @@ function SelectProvider($$interimElementProvider) {
...
+      function detachAndClean() {
+        configureAria(opts.target, false);
+
+        element.attr('opacity', 0);
+        element.removeClass('md-active');
+        detachElement(element, opts);
+
+        announceClosed(opts);
+
+        if (!opts.$destroy && opts.restoreFocus) opts.target.focus();
```
Comment (jelbourn) [inline src/components/select/select.js]: Braces

File: src/components/select/select.spec.js [hunk 11]
```diff
@@ -156,6 +55,23 @@ describe('<md-select>', function() {
     expect(called).toBe(true);
   }));
 
+  iit('closes the menu during scope.$destroy()', inject(function($document, $rootScope, $timeout) {
```
Comment (jelbourn) [inline src/components/select/select.spec.js]: iit

File: src/components/select/select.spec.js [hunk 12]
```diff
@@ -156,6 +55,23 @@ describe('<md-select>', function() {
...
+    var container = angular.element("<div></div>");
+    var scope = $rootScope.$new();
+    var select = setupSelect(' ng-model="val" ', [1, 2, 3], false, scope).find('md-select');
+
+      $document[0].body.appendChild(container[0]);
+      container.append(select);
+
+      openSelect(select);
```
Comment (jelbourn) [inline src/components/select/select.spec.js]: Inconsistent indentation

File: src/core/services/interimElement/interimElement.js [hunk 13]
```diff
@@ -331,19 +356,30 @@ function InterimElementProvider() {
...
+            return reason;
           });
 
         return interim.deferred.promise;
       }
 
+      /*
+       * Special method to quick-remove the interim element without animations
+       */
+      function $destroy() {
```
Comment (jelbourn) [inline src/core/services/interimElement/interimElement.js]: Why start with a `$`? None of the other methods do.

**Comment by topherfangio** [comment]
I signed it!

File: src/components/select/select.js [hunk 14]
```diff
@@ -316,20 +315,23 @@ function SelectDirective($mdSelect, $mdUtil, $mdTheming, $mdAria, $rootElement,
       element.attr(ariaAttrs);
 
       scope.$on('$destroy', function() {
-        if (isOpen) {
-          $mdSelect.hide().finally(function() {
```
Comment (topherfangio) [inline src/components/select/select.js]: Do we know why the `hide()` used to be called? Did we lose an animation by skipping this, or was it irrelevant?

File: src/components/select/select.js [hunk 15]
```diff
@@ -782,38 +785,40 @@ function SelectProvider($$interimElementProvider) {
      * Interim-element onRemove logic....
      */
     function onRemove(scope, element, opts) {
+      opts = opts || { };
```
Comment (topherfangio) [inline src/components/select/select.js]: If the user fails to pass options (or passes null), won't the next three lines fail because we've set it to an empty object which does not have those methods?

File: src/components/dialog/dialog.js (hunk 4 above)
Comment (ThomasBurleson) [inline src/components/dialog/dialog.js]: This is a deprecated use will be cleaned up.

File: src/components/dialog/dialog.js (hunk 5 above)
Comment (ThomasBurleson) [inline src/components/dialog/dialog.js]: `options.$destroy` is a flag related to the `$destroy` event.
we will use `if (options.$destroy === true)`
>  method names has been renamed to `destroy( ... )`

File: src/components/dialog/dialog.js (hunk 6 above)
Comment (ThomasBurleson) [inline src/components/dialog/dialog.js]: removing interim class or attribute changes. **Clean** up of interim changes.

File: src/components/dialog/dialog.js (hunk 8 above)
Comment (ThomasBurleson) [inline src/components/dialog/dialog.js]: should be `options.$destroy`; used to indicate a **skip** of hide backdrop animations.

File: src/components/dialog/dialog.spec.js (hunk 9 above)
Comment (ThomasBurleson) [inline src/components/dialog/dialog.spec.js]: done.

File: src/components/dialog/dialog.js (hunk 8 above)
Comment (jelbourn) [inline src/components/dialog/dialog.js]: Why not `options.skipBackdropCloseAnimation` ?

File: src/core/services/interimElement/interimElement.js (hunk 13 above)
Comment (ThomasBurleson) [inline src/core/services/interimElement/interimElement.js]: changed/fixed.

File: src/components/bottomSheet/demoBasicUsage/script.js [hunk 16]
```diff
@@ -22,7 +22,7 @@ angular.module('bottomSheetDemo1', ['ngMaterial'])
       controller: 'ListBottomSheetCtrl',
       targetEvent: $event
     }).then(function(clickedItem) {
-      $scope.alert = clickedItem.name + ' clicked!';
+      $scope.alert = clickedItem['name'] + ' clicked!';
```
Comment (robertmesserle) [inline src/components/bottomSheet/demoBasicUsage/script.js]: Is there a reason for this change?

File: src/components/dialog/dialog.js [hunk 17]
```diff
@@ -661,10 +685,12 @@ function MdDialogProvider($$interimElementProvider) {
...
         if (options.backdrop) {
-          $animate.leave(options.backdrop);
+          if ( !!$destroy ) options.backdrop.remove();
```
Comment (robertmesserle) [inline src/components/dialog/dialog.js]: `!!` isn't necessary

File: src/components/select/select.js (hunk 10 above)
Comment (robertmesserle) [inline src/components/select/select.js]: Do we want to make braces a requirement for all blocks?  If so, I will have to make this change in a lot of my code.

File: src/core/services/interimElement/interimElement.js [hunk 18]
```diff
@@ -190,6 +198,9 @@ function InterimElementProvider() {
 
       return publicService;
 
+      /**
+       *
+       */
```
Comment (robertmesserle) [inline src/core/services/interimElement/interimElement.js]: Eh?

File: src/core/services/interimElement/interimElement.js [hunk 19]
```diff
@@ -259,7 +282,7 @@ function InterimElementProvider() {
        */
       function show(options) {
         options = options || {};
-        var interimElement = new InterimElement(options);
+        var interimElement = new InterimElement(options || {});
```
Comment (robertmesserle) [inline src/core/services/interimElement/interimElement.js]: This is redundant with the above line `options = options || {};`

File: src/core/services/interimElement/interimElement.js [hunk 20]
```diff
@@ -311,9 +335,10 @@ function InterimElementProvider() {
 
         function closeElement(interim) {
           interim
-            .remove(reason || SHOW_CLOSED, false)
+            .remove(reason || SHOW_CLOSED, false, options || { })
```
Comment (robertmesserle) [inline src/core/services/interimElement/interimElement.js]: Also seems redundant given the above line `options = options || {};`

File: src/core/services/interimElement/interimElement.js [hunk 21]
```diff
@@ -393,39 +429,44 @@ function InterimElementProvider() {
...
          * - perform optional clean up scope.
          */
-        function transitionOutAndRemove(response, isCancelled) {
+        function transitionOutAndRemove(response, isCancelled, opts) {
+          options = angular.merge(options || {}, opts || {});
           options.cancelAutoHide && options.cancelAutoHide();
+          options.element.triggerHandler('$mdInterimElementRemove');
 
-          return $q(function(resolve, reject){
+          if ( !!options.$destroy ) {
```
Comment (robertmesserle) [inline src/core/services/interimElement/interimElement.js]: `!!` is not necessary

File: src/components/select/select.js (hunk 14 above)
Comment (rschmukler) [inline src/components/select/select.js]: The hide was called to restore scrolling

Comment (ThomasBurleson) [inline src/components/select/select.js]: the `$mdSelect.destroy()` calls the $$interimElement.destroy which in turn calls 

```js
/*
       * Special method to quick-remove the interim element without animations
       */
      function destroy() {
        var interim = stack.shift();

        return interim ? interim.remove(SHOW_CANCELLED, false, {'$destroy':true}) :
               $q.when(SHOW_CANCELLED);
      }
```

which eventually calls `onRemove()`. None of this has changed.
